### ISSUE TITLE
What is the proper way to terminate ONNX inference session in PYTHON  ?  How do i release memory allocated to weights?   

### ISSUE THREAD

**Issue by Mahaaveer** [onnx/onnx#3293]
I am using onnxruntime to perform inference in python using:

class onnxruntime.InferenceSession(path_or_bytes, sess_options=None, providers=None, provider_options=None) 

Calling Inference session function multiple times keeps adding roughly 260 MB to the memory until RAM goes bust. i have to hard reset the system to unfreeze. i suspect it might be the weights file (.onnx model file)

What is the proper way of exiting the inference session and  release memory? Without exiting the main python program .....

**Comment by TMVector** [comment]
This is really a question for https://github.com/microsoft/onnxruntime, but I believe the inference session is destroyed when the Python object is, so you can use `del` or wait for garbage collection once the session is out of scope.

```python
sess = onnxruntime.InferenceSession("model.onnx")
# use sess
del sess
```

**Comment by Mahaaveer** [comment]
> 
> 
> This is really a question for https://github.com/microsoft/onnxruntime, but I believe the inference session is destroyed when the Python object is, so you can use `del` or wait for garbage collection once the session is out of scope.
> 
> ```python
> sess = onnxruntime.InferenceSession("model.onnx")
> # use sess
> del sess
> ```

Hey! Thanks!

I tried this earlier with garbage collection:
del sess
gc.collect()

still the memory keeps increasing. 
printing sys.sizeof(sess) is around 71bytes. 
I guess sess and weights variables are not around 250 MB

**Comment by TMVector** [comment]
The Python object `sess` will be a native object (we're using a Python binding to the C++ onnxruntime code), so you won't get able to get the full associated memory usage with `sys.sizeof`.

You might find that you are leaking memory from somewhere other than the session objects. I played around a little bit in a notebook and it looks like it works roughly as I'd expect: https://colab.research.google.com/drive/1b-UStDW80PwbL9nsTW-_FJiaO7fIjHF-?usp=sharing

**Comment by askhade** [comment]
@TMVector thanks for a adding the explanation.
Closing this issue here. @Mahaaveer if you still have any concerns please open an issue in the onnxruntime repo

**Comment by Mahaaveer** [comment]
Thanks @TMVector for your explanation. It was really helpful. i was able  to minimize the memory leak by calling the session only once with program start  and calling another function to perform detection. Its not perfect but it brought down memory leak from 250 MB to 60 MB.
